opal/class/opal_include_list.c

```diff
@@ -0,0 +1,146 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil -*- */
+/*
+ * Copyright (c) 2024      Google, LLC. All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include <regex.h>
+
+#include "opal_config.h"
+
+#include "opal/class/opal_include_list.h"
+#include "opal/class/opal_object.h"
+#include "opal/include/opal/constants.h"
+#include "opal/mca/base/base.h"
+#include "opal/util/argv.h"
+#include "opal/util/output.h"
+#include "opal/util/printf.h"
+
+static void include_list_destructor(opal_include_list_t *p);
+
+static int opal_include_list_deserialize(opal_include_list_t *object, const char *value)
+{
+    /* release any current value and set to defaults */
+    include_list_destructor(object);
+
+    if (NULL == value || 0 == strlen(value)) {
+        return OPAL_SUCCESS;
+    }
+
+    if ('^' == value[0]) {
+        object->is_exclude = true;
+        ++value;
+    }
+
+    object->items = opal_argv_split(value, ',');
+    return OPAL_SUCCESS;
+}
+
+static char *opal_include_list_serialize(const opal_include_list_t *object)
+{
+    if (NULL == object->items) {
+        return strdup("");
+    }
+
+    char *tmp = opal_argv_join(object->items, ',');
+    if (object->is_exclude) {
+        char *value_str = NULL;
+        (void) opal_asprintf(&value_str, "^%s", tmp);
+        free(tmp);
+        return value_str;
+    }
+
+    return tmp;
+}
+
+static bool opal_include_list_is_set(const opal_include_list_t *object)
+{
+    return (object->items != NULL);
+}
+
+static void include_list_contructor(opal_include_list_t *p)
+{
+    p->super.deserialize = (opal_serializable_deserialize_fn_t)opal_include_list_deserialize;
+    p->super.serialize = (opal_serializable_serialize_fn_t)opal_include_list_serialize;
+    p->super.is_set = (opal_serializable_is_set_fn_t)opal_include_list_is_set;
+    p->items = NULL;
+    p->is_exclude = false;
+}
+
+static void include_list_destructor(opal_include_list_t *p)
+{
+    opal_argv_free(p->items);
+    include_list_contructor(p);
+}
+
+OBJ_CLASS_INSTANCE(opal_include_list_t, opal_object_t, include_list_contructor,
+                   include_list_destructor);
+
+static int include_list_match_regex(opal_include_list_t *include_list, const char *value,
```

Comment:
This function is doing the opposite of one might expect by reading the function name: it is not finding an item matching a regular expression from a list, but instead it is finding if an item matches one of the regular expressions in the list. At some point these two outcomes can be considered similar but their complexity is drastically different.

Assuming there is a real need for such the capability to match regular expressions stored in a list, this function need to be documented. And maybe renamed `include_regex_list_match`?

Reply:
I have a use case for this. With btl/uct we agreed on an include list of the supported memory domains. Problem is the naming is not necessarily simple. Right now with have mlx5_0 but it will fail to match other HCAs in the system. Same is true for our hardware (irdma[0-9]). I could just enumerate all the possibilities but having it be a regex makes the include list cleaner and easier to maintain (I can just put `irdma.*` in the list).